### PR TITLE
Use a standard recipe model to install nexus3

### PR DESCRIPTION
**Pull request by olive42**
- This uses a standard recipe with `ark' doing most of the job.
- Update unit tests accordingly.

Things to check in review:
- serverspec tests will fail because some configuration files trigger a
  service restart, which happens asynchronously at the end of the recipe run,
  and right during the tests run.
- I am still not sure the Windows recipe and tests are ok.

### CODE REVIEW

File: recipes/default.rb [hunk 1]
```diff
@@ -1,3 +1,160 @@
...
+## Install Unix service
+def systype
+  return 'windows' if platform?('windows')
+  return 'systemd' if ::File.exist?('/proc/1/comm') && ::File.open('/proc/1/comm').gets.chomp == 'systemd'
+  'sysvinit'
+end
+
+case systype
+when 'windows'
+  batch 'install Windows service' do
```
Comment (Annih) [inline recipes/default.rb]: You already did that L45

File: spec/support/matchers.rb [hunk 2]
```diff
@@ -0,0 +1,3 @@
+def put_ark(resource_name)
```
Comment (Annih) [inline spec/support/matchers.rb]: Do you really need this? Chefspec should already define it for you

Comment (olive42) [inline spec/support/matchers.rb]: Hmmm, for a moment the tests didn't work, that's why I had added this.

File: libraries/nexus_helper.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
+require 'open-uri'
+require 'timeout'
+
+module Nexus3
+  # Helper library for testing Nexus3 responses
+  module Helper
+    class Nexus3Timeout < Timeout::Error; end
```
Comment (olive42) [inline libraries/nexus_helper.rb]: Not necessarily.

File: libraries/nexus_helper.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        begin
+          open(url)
+        rescue SocketError,
+               Errno::ECONNREFUSED,
+               Errno::ECONNRESET,
+               Errno::ENETUNREACH,
+               OpenURI::HTTPError => e
+          # Getting 403 is ok since it means we reached the endpoint and
+          # it's asking us for authentication.
+          return if e.message =~ /^403/
```
Comment (olive42) [inline libraries/nexus_helper.rb]: ok

File: recipes/default.rb [hunk 5]
```diff
@@ -1,3 +1,160 @@
...
+# Fetch and un{tar,zip} Nexus3 archive
+ark "nexus-#{node['nexus3']['version']}" do
+  action :put
+  url node['nexus3']['url']
+  checksum node['nexus3']['checksum']
+  path node['nexus3']['path']
+  owner usr
+  group grp
+
+  notifies(:run, 'batch[install Nexus service]', :immediately) if platform?('windows')
```
Comment (olive42) [inline recipes/default.rb]: Changed to use 'sc query ...'

File: recipes/default.rb [hunk 6]
```diff
@@ -1,3 +1,160 @@
...
+end
+
+if platform?('windows')
+  batch 'install Nexus service' do
```
Comment (olive42) [inline recipes/default.rb]: Updated.

File: recipes/default.rb (hunk 1 above)
Comment (olive42) [inline recipes/default.rb]: Updated.

File: libraries/matchers.rb [hunk 7]
```diff
@@ -1,10 +1,6 @@
 if defined?(ChefSpec)
-  def install_nexus3(resource_name)
-    ChefSpec::Matchers::ResourceMatcher.new(:nexus3, :install, resource_name)
-  end
-
-  def uninstall_nexus3(resource_name)
-    ChefSpec::Matchers::ResourceMatcher.new(:nexus3, :uninstall, resource_name)
+  def put_ark(resource_name)
```
Comment (olive42) [inline libraries/matchers.rb]: FIxed.

**Comment by olive42** [comment]
1. I need to find a clean way to figure out if the system is using systemd or sysvinit or something else (checking on /proc/1/comm yields incoherent results in Docker tests depending on the host machine)
2. Installation needs to be a resource as discussed with @jmauro 

This is likely the wrong solution.

**Comment by olive42** [comment]
I will likely close this PR once the next one is done.

**Pull request closed without merging**